---
title: "3_mMAB_characterization"
author: "Jordi Camps"
date: "7 January 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load packages
```{r message=FALSE}
library(readxl)
library(tidyr)
library(dplyr)
library(ggplot2)
library(qpcrviia7)
library(ggpubr)
```

## Load data
```{r}
qpcr <- read_excel("data/fig3-4_qPCR_cells/2016-04-08 174453 JC TP Muscle progenitors characterization-ViiA7-export.xlsx", sheet = 1, col_names = TRUE, skip = 35)
head(qpcr)
```

## Create celltype and mouse column
```{r}
qpcr <- separate(qpcr, Sample_Name, c("temp1", "temp2"), sep = " ")
qpcr$Celltype <- qpcr$temp1
qpcr$Mouse <- qpcr$temp2
qpcr <- unite(qpcr, Sample, c("temp1", "temp2"), sep = " ")
head(qpcr)
```

## Create genotype column
```{r}
temp <- data_frame(Mouse = c("SGCB1", "SGCB2", "SGCB3", "WT8/9", "WT10/11", "WT12/13", "C3H"),
                   Genotype = c("Dystrophic", "Dystrophic", "Dystrophic", "Healthy", "Healthy", "Healthy", "C2C12"))
qpcr <- qpcr %>%
  left_join(temp)
head(qpcr)
```

## Creating numeric CT column and change undetermined values into 40
```{r}
qpcr$CT <- as.numeric(qpcr$CT)
qpcr[is.na(qpcr$CT), "CT"] <- 40
```

## Quality control
### Samples
```{r fig.height=4, fig.width=6}
ggplot(qpcr, aes(Sample, CT)) +
  geom_boxplot() +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.4, hjust = 1)) +
  ggtitle("CT values per sample")
```

### Genes
```{r fig.height=4, fig.width=6}
ggplot(qpcr, aes(Target_Name, CT)) +
  geom_boxplot() +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.4, hjust = 1)) +
  ggtitle("CT values per primer")
``` 

## Remove blanc
```{r}
qpcr <- filter(qpcr, Celltype != "Blanc")
```

## Check HKG over samples
```{r}
qpcr %>%
  filter(Target_Name == "Psma3" | Target_Name == "Rab35" | Target_Name == "Rpl13a") %>%
  ggplot(aes(y = CT, x = Celltype)) +
    geom_boxplot() +
    facet_wrap(~Target_Name, scales = "free_y")
```

## Normalize to HKG
normalize_hkg function
```{r}
normalize_hkg <- function(df, hkg, sample_col = "Sample Name", target_col = "Target Name") {
  temp <- NULL
  temp2 <- NULL
  #create df with average hkg per sample
  temp <- df[which(df[[target_col]] %in% hkg), ]
  temp <- temp %>%
    select_(sample_col, "CT") %>%
    group_by_(sample_col) %>%
    summarize(CT_hkg = gm_mean(CT, na.rm = TRUE))
  #add avg hkg to df and calculate delta ct and rel expr
  temp2 <- df[-which(df[[target_col]] %in% hkg), ]
  print(temp2 %>%
    group_by_(sample_col) %>%
    left_join(temp) %>%
    mutate(Delta_ct = CT_hkg - CT, Rel_expr = 2^Delta_ct)
    )
}
```

gm_mean function
```{r}
gm_mean = function(x, na.rm=TRUE){
  exp(sum(log(x[x > 0]), na.rm=na.rm) / length(x))
}
```

```{r}
qpcr <- normalize_hkg(df = qpcr, hkg = "Rpl13a", sample_col = "Sample", target_col = "Target_Name")
```

## Plot
```{r fig.height=5, fig.width=10}
#my_comparisons <- list(c("Alpl+/Sca1+", "Sca1+"))
qpcr$Celltype <- plyr::revalue(qpcr$Celltype, c(C2C12_nLacZ = "C2C12_nLacZ", `Alpl+` = "Alpl+", `Alpl+/Sca1+` = "MAB", `Sca1+` = "FAP"))
qpcr$Celltype <- factor(qpcr$Celltype, levels = c("FAP", "MAB", "Alpl+", "C2C12_nLacZ"))
qpcr %>%
  filter(Target_Name != "Psma3" & Target_Name != "Rab35" & Target_Name != "Mrf4" & Target_Name != "Myh1" & Target_Name != "Pax3") %>%
  filter(Celltype != "C2C12_nLacZ" & Celltype != "Alpl+") %>%
  ggplot(aes(x = Celltype, y = Rel_expr, col = Celltype)) +
  #geom_dotplot(binaxis = "y", stackdir = "center", dotsize = 3) +
  #stat_summary(fun.y = mean, fun.ymax = mean, fun.ymin = mean, col = "black", geom = "Crossbar") +
  geom_boxplot(outlier.shape = NA, size = .8) +
  geom_jitter(width = 0.3) +
  facet_wrap(~Target_Name, scales = "free", nrow = 3) +
  scale_y_continuous("Relative expression to Rpl13a") +
  stat_compare_means(method = "t.test", label = "p.signif") +
  ggthemes::scale_color_tableau() +
  theme_bw(base_size = 16) +
  theme(panel.border = element_blank(), panel.grid = element_blank(), axis.line = element_line(colour = "black"), 
        axis.ticks.x = element_blank(), axis.ticks.y = element_line(colour = "black"), axis.text.x = element_blank(), 
        axis.text.y = element_text(colour = "black", size = 8), axis.title.x = element_blank(), strip.background = element_blank(),
        strip.text = element_text(face = "italic"), legend.title = element_blank(), legend.position = "bottom", 
        plot.margin = unit(c(0, 0, 0, 0), "cm"))
ggsave("plots/cells/mMAB_characterization.pdf", dpi = 600)
ggsave("plots/cells/mMAB_characteriation.png", dpi = 600)
```